### PR TITLE
DeepLabV3, DeepLabV3+ example based on MindCV

### PR DESCRIPTION
**Pull request by Fzilan**
Thank you for your contribution to the MindCV repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

DeepLabV3 example based on MindCV

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

**Comment by Fzilan** [comment]
deeplabv3_os16

File: examples/seg/deeplabv3/README.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+2. Install dependencies as shown [here](https://mindspore-lab.github.io/mindcv/installation/)
+
+3. Prepare dataset
+
+- Download Pascal VOC 2012 dataset,  [VOC2012](http://host.robots.ox.ac.uk/pascal/VOC/voc2012/) and Semantic Boundaries Dataset, [SBD](https://www2.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/semantic_contours/benchmark.tgz).
+
+- Prepare training and test data list files with the path to image and annotation pairs. You could simply run `python examples/seg/deeplabv3/build_dataset/get_dataset_lst.py --data_root=[PATH TO DATA]` to generate the list files. The lines in list file should be like as follows:
+
+  ```
+  /[PATH..]/JPEGImages/2007_000032.jpg /[PATH..]/SegmentationClassGray/2007_000032.png
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/README.md:50]: what's `[PATH..]`?
tips: use `/path/to/data` or shell var like `${DATA_ROOT}`

File: examples/seg/deeplabv3/README.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+### Train
+
+Specify `deeplabv3`  or  `deeplabv3plus` in key word `model` in the config file.
+
+It is highly recommended to use **distributed training** for this DeepLabV3 and DeepLabV3+ implementation.
+
+For distributed training using **OpenMPI's `mpirun`**, simply run
+```shell
+cd mindcv  # change directory to the root of MindCV repository
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/README.md:79]: hi, didn't we in the root of project when preparing data? Assumption like "We are always in the `${PROJECT_ROOT}`" may be better

File: examples/seg/deeplabv3/README.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+for ((i = 0; i < ${DEVICE_NUM}; i++)); do
+   export DEVICE_ID=$i
+   export RANK_ID=$i
+   python -u examples/seg/deeplabv3/train.py --config [the path to the config file]  &> ./train_$i.log &
+done
+```
+
+and start training by running:
+```shell l
+cd mindcv  # change directory to the root of MindCV repository
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/README.md:100]: same as above

File: examples/seg/deeplabv3/README.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+bash ascend8p.sh
+```
+
+For single-device training, simply set the parameter ``distributed`` to ``False`` in config file and run:
+```shell
+cd mindcv  # change directory to the root of MindCV repository
+python examples/seg/deeplabv3/train.py --config [the path to the config file]
+```
+
+##### The training steps are as follow:
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/README.md:110]: whats the difference from the above code snippet?

File: examples/seg/deeplabv3/README.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  # for deeplabv3+
+  mpirun -n 8 python examples/seg/deeplabv3/train.py --config examples/seg/deeplabv3/deeplabv3plus_s8_dilated_resnet101.yaml
+  ```
+
+### Test
+
+For testing the trained model, first specify the path to the model checkpoint at keyword `ckpt_path` in the config file. You could modify `output_stride`, `flip`, `scales` in config file during inference.
+```shell
+cd mindcv  # change directory to the root of MindCV repository
+python examples/seg/deeplabv3/eval.py --config [the path to the config file]
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/README.md:137]: remove this one

File: examples/seg/deeplabv3/bulid_dataset/bulid_seg_data.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import argparse
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/bulid_dataset/bulid_seg_data.py:1]: files in `bulid_dataset`  are all about preprocess, right? need a better name.
btw, "build", not "bulid".

File: examples/seg/deeplabv3/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
+import os
+import stat
+
+from postprocess import apply_eval
+
+from mindspore import log as logger
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/callbacks.py:6]: DO NOT USE `mindspore.log`, refer to `mindcv.utils.callback`

File: examples/seg/deeplabv3/callbacks.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            logger.warning("ValueError, failed to remove the older ckpt file %s.", file_name)
+
+    def on_train_epoch_end(self, run_context):
+        """Callback when epoch end."""
+        cb_params = run_context.original_args()
+        cur_epoch = cb_params.cur_epoch_num
+
+        if cur_epoch >= self.eval_start_epoch and (cur_epoch - self.eval_start_epoch) % self.interval == 0:
+            res = self.eval_function(self.eval_param_dict)
+            print("epoch: {}, {}: {}".format(cur_epoch, self.metrics_name, res), flush=True)
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/callbacks.py:75]: use `logging` instead of `print`

File: examples/seg/deeplabv3/config/deeplabv3_s16_dilated_resnet101.yaml [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+mode: 1
+distribute: True
+num_parallel_workers: 8
+enable_modelarts: False
+device_target: "Ascend"
+all_reduce_fusion_config: [90, 183, 279]
+
+# dataset
+dataset: "vocaug"
+data_dir: "/home/cvgroup/Datasets/Segment/mindrecord/vocaug0"
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/config/deeplabv3_s16_dilated_resnet101.yaml:14]: /path/to

File: examples/seg/deeplabv3/config/deeplabv3_s16_dilated_resnet101.yaml [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+image_std: [57.375, 57.120, 58.395]
+max_scale: 2.0
+min_scale: 0.5
+ignore_label: 255
+num_classes: 21
+shuffle: True
+
+# backbone
+backbone: "dilated_resnet101"
+backbone_ckpt_path: "/home/cvgroup/fzl/checkpoint/resnet/resnet101-689c5e77.ckpt"
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/config/deeplabv3_s16_dilated_resnet101.yaml:27]: /path/to

File: examples/seg/deeplabv3/deeplabv3.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        x = ops.ResizeNearestNeighbor(size=(size[2], size[3]), align_corners=True)(x)
+        x = ops.cat((x, low_level_features), axis=1)
+        x = self.last_conv(x)
+
+        return x
+
+
+class DeeplabV3Plus(nn.Cell):
+    """
+    DeepLabV3Plus implement.
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/deeplabv3.py:210]: annotations for main model classes should be more detailed. Please refer to other models

File: examples/seg/deeplabv3/deeplabv3.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+    def construct(self, x):
+        size = ops.shape(x)
+        low_level_features, features = self.backbone(x)
+        out = self.aspp(features)
+        out = self.decoder(out, low_level_features)
+        out = ops.interpolate(out, size=(size[2], size[3]), mode="bilinear", align_corners=True)
+        return out
+
+
+class BuildInferNetwork(nn.Cell):
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/deeplabv3.py:240]: Weird name. and why we need it

File: examples/seg/deeplabv3/deeplabv3.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    """
+    Provide infer network.
+
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/deeplabv3.py:243]: redundant empty line

File: examples/seg/deeplabv3/train.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
+import argparse
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/train.py:1]: setup logging in this file, and use it everywhere.

File: examples/seg/deeplabv3/train.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+if __name__ == "__main__":
+    args = parse_args()
+    yaml_fp = args.config
+    with open(yaml_fp) as fp:
+        args = yaml.safe_load(fp)
+    args = Dict(args)
+    context.set_context(device_target=args.device_target)
+    # data sync for cloud platform if enabled
+
+    if args.enable_modelarts:
```
Comment (geniuspatrick) [inline examples/seg/deeplabv3/train.py:179]: remove this

File: examples/seg/deeplabv3/deeplabv3.py (hunk 12 above)
Comment (Fzilan) [reply examples/seg/deeplabv3/deeplabv3.py:240]: rename to DeepLabInferNetwork. 
DeepLabV3 and DeepLabV3Plus share the same infer network.

File: examples/seg/deeplabv3/README.md (hunk 4 above)
Comment (Fzilan) [reply examples/seg/deeplabv3/README.md:110]: according to paper, you should finetune twice, with differnet os setting, to get the final best weights.
the above code snippet shows how to start a training here.  "The training steps"  emphasizes the training protocol based on paper, that you should first use os=16 and then os=8.

File: examples/seg/deeplabv3/bulid_dataset/bulid_seg_data.py (hunk 6 above)
Comment (Fzilan) [reply examples/seg/deeplabv3/bulid_dataset/bulid_seg_data.py:1]: `bulid_dataset ` rename to `preprocess`

**Review by geniuspatrick — APPROVED** [review]

**Review by vigo999 — APPROVED** [review]
